### PR TITLE
Upgrade to react-router 6.2.2 addressing #9913

### PR DESCRIPTION
**Pull request by testn**
Signed-off-by: Testo Nakada <test1@doramail.com>

## Hey, I just made a Pull Request!

Upgrade react-router from 6.0.0-beta to react-router 6.2.2

#### :heavy_check_mark: Checklist

- [X] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [X] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: .changeset/spotty-dogs-hug.md
Comment (freben) [inline .changeset/spotty-dogs-hug.md:2]: This package is not published and can be skipped here.


Also no breaking changes are 'patch' now that we are at 0.x versions

Comment (testn) [reply .changeset/spotty-dogs-hug.md:2]: Ok. BTW, @freben  can you help me check the weird error?

```
packages/app/src/components/Root/Root.tsx:104:12 - error TS2322: Type '{ icon: OverridableComponent<SvgIconTypeMap<{}, "svg">>; to: string; text: string; }' is not assignable to type 'IntrinsicAttributes & SidebarItemProps'.
  Property 'children' is missing in type '{ icon: OverridableComponent<SvgIconTypeMap<{}, "svg">>; to: string; text: string; }' but required in type '{ to?: string | undefined; onClick?: ((ev: MouseEvent<Element, MouseEvent>) => void) | undefined; children: ReactNode; }'.

104           <SidebarItem
               ~~~~~~~~~~~
```

I couldn't find the changes to libraries other than the ones I changed. But it broke `yarn tsc`

Comment (testn) [reply .changeset/spotty-dogs-hug.md:2]: It sounds similar to https://stackoverflow.com/questions/61902377/react-typescript-error-for-material-ui-icons

Comment (freben) [reply .changeset/spotty-dogs-hug.md]: Essentially i think you'll find that there will be pretty significant work in making tsc happy. That's why we deferred this and asked for community help since we didn't have time to thoroughly work through it at this point.

Comment (testn) [reply .changeset/spotty-dogs-hug.md]: Let me figure out how to fix that.

Comment (testn) [reply .changeset/spotty-dogs-hug.md]: Indeed, it is such a pain to fix it. react-router made a lot of breaking changes before it went out of beta release.

Comment (testn) [reply .changeset/spotty-dogs-hug.md]: @freben I managed to fix the problem. However, I'm not sure about NavLink one. Can you help see whether I did it correctly?

File: packages/core-components/src/layout/Sidebar/Items.tsx
Comment (freben) [inline packages/core-components/src/layout/Sidebar/Items.tsx:294]: @Rugvip some eyes on this one

File: plugins/tech-insights/src/components/ScorecardsOverview/ScorecardsOverview.tsx
Comment (freben) [inline plugins/tech-insights/src/components/ScorecardsOverview/ScorecardsOverview.tsx:35]: Ah this one was bad to begin with. Could you remove the `useParams` hook usage here and instead do `const { entity } = useEntity()` from the package `@backstage/plugin-catalog-react`? Then you can pass in `getCompoundEntityRef(entity)` from `@backstage/catalog-model` below.

File: plugins/techdocs/src/reader/components/LegacyTechDocsPage.tsx
Comment (freben) [inline plugins/techdocs/src/reader/components/LegacyTechDocsPage.tsx:33]: could you name this one `entityRef` instead of `doc`?

File: plugins/techdocs/src/reader/components/TechDocsReaderPage.tsx
Comment (freben) [inline plugins/techdocs/src/reader/components/TechDocsReaderPage.tsx:60]: could you name these `entityRef` instead of `doc`?

File: yarn.lock
Comment (freben) [inline yarn.lock:20966]: @Rugvip @benjdlambert ALL of the references to the old beta have to be gone from the lock file before this can be merged right? Or is there any way that this can be done piecemeal? I think these come from roadie plugins and similar

**Review by freben — COMMENTED** [review]
Thanks! Some comments and also pinged colleagues for a couple things

File: plugins/cloudbuild/src/components/WorkflowRunDetails/useWorkflowRunsDetails.ts
Comment (testn) [reply plugins/cloudbuild/src/components/WorkflowRunDetails/useWorkflowRunsDetails.ts:28]: Yeah, I'm confused what I should do with that, too.

File: plugins/scaffolder/src/components/TemplatePage/TemplatePage.tsx
Comment (testn) [reply plugins/scaffolder/src/components/TemplatePage/TemplatePage.tsx:122]: Are you saying that it is fine not to validate that the param is provided?

File: plugins/techdocs/src/reader/components/TechDocsReaderPage.tsx
Comment (testn) [reply plugins/techdocs/src/reader/components/TechDocsReaderPage.tsx:60]: definitely!

File: packages/core-components/src/layout/Sidebar/Items.tsx
Comment (Rugvip) [reply packages/core-components/src/layout/Sidebar/Items.tsx:294]: Yep this should now be fixed, best to verify though

File: yarn.lock
Comment (Rugvip) [reply yarn.lock:20966]: I'd say we need to figure out how to do this piecemeal. Forcing everyone to migrate to v6 stable immediately isn't really something I think we should push out unless there are no other options.

**Review by Rugvip — CHANGES_REQUESTED** [review]
These are all changes in the right direction and thank you for doing this work! 👍

Gonna block the merge of this for now thought, as I want us to explore our options for providing a smoother migration. This would currently push out a big bang that forces everyone to migrate at once, and where existing plugins that have been published break.

I'm not exactly sure how and if that can be done, but a starting point to explore is whether we can provide the React context for both versions of react-router at once. It's possible to install separate versions of the same dependency like [this](https://github.com/backstage/backstage/blob/1d9a123dde2b655400c5b9fb6ebc334cee3d5038/packages/cli/package.json#L40-L41)

File: plugins/tech-insights/src/components/ScorecardsOverview/ScorecardsOverview.tsx
Comment (testn) [reply plugins/tech-insights/src/components/ScorecardsOverview/ScorecardsOverview.tsx]: Ahhhhh. Thanks! That is much better.

**Comment by testn** [comment]
> These are all changes in the right direction and thank you for doing this work! 👍
> 
> Gonna block the merge of this for now thought, as I want us to explore our options for providing a smoother migration. This would currently push out a big bang that forces everyone to migrate at once, and where existing plugins that have been published break.
> 
> I'm not exactly sure how and if that can be done, but a starting point to explore is whether we can provide the React context for both versions of react-router at once. It's possible to install separate versions of the same dependency like [this](https://github.com/backstage/backstage/blob/1d9a123dde2b655400c5b9fb6ebc334cee3d5038/packages/cli/package.json#L40-L41)

Guess it can not be done easily as different versions have different apis. How can we achieve something like that?

File: plugins/tech-insights/src/components/ScorecardsOverview/ScorecardsOverview.tsx
Comment (freben) [reply plugins/tech-insights/src/components/ScorecardsOverview/ScorecardsOverview.tsx]: Actually no. In fact, these components have no business trying to parse the url at all. The way we've set things up is that there is an entity context around the entire entity page where this should be placed. That way it only has to be loaded once and everything can access that instance freely

**Comment by Rugvip** [comment]
@testn Not sure. @benjdlambert had the idea of doing something similar to what we did for React itself. Basically making react-router a peer dependency and allowing both the beta and stable version in the version range, which makes it possible to choose. The problem with that is that any API incompatibilities between v6 and the beta will horribly explode the app, so it would be a sort of smooth-ish migration that will let most thing function short-term but it would still be heavily encouraged to move over to v6 asap. Whether that's viable depends largely on what kind of breakages there are in the API, and if it's going to be common to encounter them in plugins outside of this repo.

**Pull request closed without merging**
